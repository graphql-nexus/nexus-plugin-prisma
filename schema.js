module.exports = require('./dist/schema-entrypoint')
