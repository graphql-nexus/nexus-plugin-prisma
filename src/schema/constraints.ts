import { DmmfDocument } from './dmmf'

/**
 * Find the unique identifiers necessary to indentify a field
 *
 * Unique fields for a model can be one of (in this order):
 * 1. One (and only one) field with an @id annotation
 * 2. Multiple fields with @@id clause
 * 3. One (and only one) field with a @unique annotation (if there are multiple, use the first one)
 * 4. Multiple fields with a @@unique clause
 */
export function resolveUniqueIdentifiers(typeName: string, dmmf: DmmfDocument): string[] {
  const model = dmmf.getModelOrThrow(typeName)

  // Try finding 1.
  const singleIdField = model.fields.find((f) => f.isId)

  if (singleIdField) {
    return [singleIdField.name]
  }

  // Try finding 2.
  if (model.idFields && model.idFields.length > 0) {
    return model.idFields
  }

  const singleUniqueField = model.fields.find((f) => f.isUnique)

  if (singleUniqueField) {
    return [singleUniqueField.name]
  }

  if (model.uniqueFields && model.uniqueFields.length > 0) {
    return model.uniqueFields[0]
  }

  throw new Error(`Unable to resolve a unique identifier for the Prisma model: ${model.name}`)
}

export function findMissingUniqueIdentifiers(
  data: Record<string, any>,
  uniqueIdentifiers: string[]
): string[] | null {
  const missingIdentifiers: string[] = []

  for (const identifier of uniqueIdentifiers) {
    if (!data[identifier]) {
      missingIdentifiers.push(identifier)
    }
  }

  if (missingIdentifiers.length > 0) {
    return missingIdentifiers
  }

  return null
}

export function buildWhereUniqueInput(data: Record<string, any>, uniqueIdentifiers: string[]) {
  if (uniqueIdentifiers.length === 1) {
    return pickFromRecord(data, uniqueIdentifiers)
  }

  const compoundName = uniqueIdentifiers.join('_')

  return {
    [compoundName]: pickFromRecord(data, uniqueIdentifiers),
  }
}

function pickFromRecord(record: Record<string, any>, keys: string[]) {
  const output: Record<string, any> = {}

  for (const identifier of keys) {
    if (record[identifier]) {
      output[identifier] = record[identifier]
    }
  }

  return output
}
