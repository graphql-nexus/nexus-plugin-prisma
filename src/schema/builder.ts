import * as Nexus from '@nexus/schema'
import { DynamicOutputPropertyDef } from '@nexus/schema/dist/dynamicProperty'
import { defaultFieldResolver, GraphQLFieldResolver, GraphQLScalarType } from 'graphql'
import * as path from 'path'
import * as Constraints from './constraints'
import { addComputedInputs, DmmfDocument, DmmfTypes, getTransformedDmmf } from './dmmf'
import * as GraphQL from './graphql'
import {
  OnUnknownArgName,
  OnUnknownFieldName,
  OnUnknownFieldType,
  OnUnknownPrismaModelName,
  raiseErrorOrTriggerHook,
} from './hooks'
import { getCrudMappedFields } from './mapping'
import {
  ArgsNamingStrategy,
  defaultArgsNamingStrategy,
  defaultFieldNamingStrategy,
  FieldNamingStrategy,
  OperationName,
} from './naming-strategies'
import { transformNullsToUndefined } from './null'
import { paginationStrategies, PaginationStrategyTypes } from './pagination'
import { proxifyModelFunction, proxifyPublishers } from './proxifier'
import { Publisher } from './publisher'
import * as Typegen from './typegen'
import {
  assertPrismaClientInContext,
  GlobalComputedInputs,
  Index,
  indexBy,
  isEmptyObject,
  LocalComputedInputs,
  lowerFirst,
} from './utils'

interface FieldPublisherConfig {
  alias?: string
  type?: Nexus.core.AllOutputTypes
  pagination?: boolean | Record<string, boolean>
  filtering?: boolean | Record<string, boolean>
  ordering?: boolean | Record<string, boolean>
  computedInputs?: LocalComputedInputs<any>
  resolve?: (
    root: object,
    args: object,
    ctx: object,
    info: object,
    originalResolve: GraphQLFieldResolver<any, any, any>
  ) => Promise<any>
}

type WithRequiredKeys<T, K extends keyof T> = Omit<T, K> & Required<Pick<T, K>>
// Config options that are populated with defaults will not be undefined
type ResolvedFieldPublisherConfig = Omit<
  WithRequiredKeys<FieldPublisherConfig, 'alias' | 'type'>,
  'computedInputs'
  // Internally rename the arg passed to a resolver as 'computedInputs' to clarify scope
> & { locallyComputedInputs: LocalComputedInputs<any> }

type FieldPublisher = (opts?: FieldPublisherConfig) => PublisherMethods // Fluent API
type PublisherMethods = Record<string, FieldPublisher>
type PublisherConfigData = {
  field: DmmfTypes.SchemaField
  givenConfig?: FieldPublisherConfig
}
type FieldConfigData = {
  field: DmmfTypes.SchemaField
  publisherConfig: ResolvedFieldPublisherConfig
  typeName: string
  operation?: OperationName | null
  resolve?: Nexus.FieldResolver<any, string>
}

/**
 * When dealing with list types we rely on the list type zero value (empty-list)
 * to represent the idea of null.
 *
 * For Prisma Client JS' part, it will never return null for list type fields nor will it
 * ever return null value list members.
 */
const dmmfListFieldTypeToNexus = (fieldType: DmmfTypes.SchemaField['outputType']) => {
  return fieldType.isList
    ? {
        list: [true],
        nullable: false,
      }
    : {
        nullable: !fieldType.isRequired,
      }
}

type PrismaClientFetcher = (ctx: Nexus.core.GetGen<'context'>) => any

export interface Options {
  // TODO return type should be Prisma Client
  /**
   * nexus-prisma will call this to get a reference to an instance of the Prisma Client.
   * The function is passed the context object. Typically a Prisma Client instance will
   * be available on the context to support your custom resolvers. Therefore the
   * default getter returns `ctx.prisma`.
   */
  prismaClient?: PrismaClientFetcher
  /**
   * Same purpose as for that used in `Nexus.makeSchema`. Follows the same rules
   * and permits the same environment variables. This configuration will completely
   * go away once Nexus has typeGen plugin support.
   */
  shouldGenerateArtifacts?: boolean
  inputs?: {
    /**
     * Where can nexus-prisma find the Prisma Client JS package? By default looks in
     * `node_modules/@prisma/client`. This is needed because nexus-prisma
     * gets your Prisma schema AST and Prisma Client JS crud info from the generated
     * Prisma Client JS package.
     */
    prismaClient?: string
  }
  outputs?: {
    /**
     * Where should nexus-prisma put its typegen on disk?
     *
     * @default 'node_modules/@types/typegen-nexus-plugin-prisma/index.d.ts'
     *
     * @remarks
     *
     * This configuration will completely go away once Nexus has typeGen plugin
     * support.
     *
     */
    typegen?: string
  }
  /**
   * Select the pagination strategy.
   *
   * 'prisma' strategy results in GraphQL pagination arguments mirroring those of Prisma: skip, cursor, take
   *
   * 'relay' strategy results in GraphQL pagination arguments matching those of the [GraphQL Relay specification](https://relay.dev/graphql/connections.htm): before, after, first, last.
   *
   * @default 'relay'
   */
  paginationStrategy?: 'relay' | 'prisma'
  /**
   * Enable experimental CRUD capabilities.
   * Add a `t.crud` method in your definition block to generate CRUD resolvers in your `Query` and `Mutation` GraphQL Object Type.
   *
   * @default false
   */
  experimentalCRUD?: boolean
  /**
   * Map of GraphQL scalar types to be used by the library for the Prisma scalars
   *
   * When not provided, the scalar types will be passthrough.
   *
   * @default {}
   */
  scalars?: Partial<Record<Typegen.GetGen<'scalars', string>, GraphQLScalarType>>
  computedInputs?: GlobalComputedInputs
}

export interface InternalOptions extends Options {
  dmmf?: DmmfDocument // For testing
  nexusBuilder: Nexus.PluginBuilderLens
  nexusPrismaImportId?: string
  onUnknownFieldName?: OnUnknownFieldName // Framework option
  onUnknownFieldType?: OnUnknownFieldType // Framework option
  onUnknownArgName?: OnUnknownArgName // Framework option
  onUnknownPrismaModelName?: OnUnknownPrismaModelName // Framework option
  initializedByFramework?: boolean // Framework option
}

export interface InternalPublicOptions extends Omit<InternalOptions, 'nexusBuilder'> {}

export function build(options: InternalOptions) {
  const builder = new SchemaBuilder(options)

  return {
    types: builder.build(),
    wasCrudUsedButDisabled() {
      return builder.wasCrudUsedButDisabled
    },
  }
}

// The @types default is based on the privileged given to such
// packages by TypeScript. For details refer to https://www.typescriptlang.org/docs/handbook/tsconfig-json.html#types-typeroots-and-types
let defaultTypegenPath: string
if (process.env.NEXUS_PRISMA_TYPEGEN_PATH) {
  defaultTypegenPath = process.env.NEXUS_PRISMA_TYPEGEN_PATH
} else if (process.env.LINK) {
  defaultTypegenPath = path.join(process.cwd(), 'node_modules/@types/typegen-nexus-plugin-prisma/index.d.ts')
} else {
  defaultTypegenPath = path.join(__dirname, '../../../@types/typegen-nexus-plugin-prisma/index.d.ts')
}

let defaultClientPath: string

if (process.env.NEXUS_PRISMA_CLIENT_PATH) {
  defaultClientPath = process.env.NEXUS_PRISMA_CLIENT_PATH
} else if (process.env.LINK) {
  defaultClientPath = path.join(process.cwd(), '/node_modules/@prisma/client')
} else {
  defaultClientPath = '@prisma/client'
}

// NOTE This will be replaced by Nexus plugins once typegen integration is available.
const shouldGenerateArtifacts =
  process.env.NEXUS_SHOULD_GENERATE_ARTIFACTS === 'true'
    ? true
    : process.env.NEXUS_SHOULD_GENERATE_ARTIFACTS === 'false'
    ? false
    : Boolean(!process.env.NODE_ENV || process.env.NODE_ENV === 'development')

const defaultOptions = {
  shouldGenerateArtifacts,
  prismaClient: (ctx: any) => ctx.prisma,
  paginationStrategy: 'relay' as const,
  inputs: {
    prismaClient: defaultClientPath,
  },
  outputs: {
    typegen: defaultTypegenPath,
  },
  computedInputs: {},
}

export interface CustomInputArg {
  arg: DmmfTypes.SchemaArg
  type: DmmfTypes.InputType | DmmfTypes.Enum | { name: string } // scalar
}

export class SchemaBuilder {
  readonly dmmf: DmmfDocument
  protected argsNamingStrategy: ArgsNamingStrategy
  protected fieldNamingStrategy: FieldNamingStrategy
  protected paginationStrategy: PaginationStrategyTypes
  protected getPrismaClient: PrismaClientFetcher
  protected publisher: Publisher
  protected scalars: Record<string, GraphQLScalarType>
  protected globallyComputedInputs: GlobalComputedInputs
  protected unknownFieldsByModel: Index<string[]>
  public wasCrudUsedButDisabled: boolean

  constructor(public options: InternalOptions) {
    const config = {
      ...defaultOptions,
      ...options,
      inputs: { ...defaultOptions.inputs, ...options.inputs },
      outputs: { ...defaultOptions.outputs, ...options.outputs },
    }
    // Internally rename the 'computedInputs' plugin option to clarify scope
    this.globallyComputedInputs = config.computedInputs ? config.computedInputs : {}
    this.paginationStrategy = paginationStrategies[config.paginationStrategy]
    this.dmmf =
      options.dmmf ||
      getTransformedDmmf(config.inputs.prismaClient, {
        globallyComputedInputs: this.globallyComputedInputs,
        paginationStrategy: this.paginationStrategy,
      })
    this.scalars = (options.scalars as any) ?? {}
    this.publisher = new Publisher(this.dmmf, config.nexusBuilder, this.scalars)
    this.unknownFieldsByModel = {}

    this.argsNamingStrategy = defaultArgsNamingStrategy
    this.fieldNamingStrategy = defaultFieldNamingStrategy
    this.wasCrudUsedButDisabled = false

    this.getPrismaClient = (ctx: any) => {
      const prismaClient = config.prismaClient(ctx)
      assertPrismaClientInContext(prismaClient)
      return prismaClient
    }
    if (config.shouldGenerateArtifacts) {
      Typegen.generateSync({
        prismaClientPath: config.inputs.prismaClient,
        typegenPath: config.outputs.typegen,
        paginationStrategy: this.paginationStrategy,
        nexusPrismaImportId: options.nexusPrismaImportId,
      })
    }
  }

  /**
   * The build entrypoint, bringing together sub-builders.
   */
  build() {
    if (this.options.experimentalCRUD === true) {
      return [this.buildCRUD(), this.buildModel()]
    }

    return [this.buildModel(), this.buildDisabledCRUD()]
  }

  /**
   * Build `t.crud` dynamic output property
   */
  protected buildCRUD(): DynamicOutputPropertyDef<'crud'> {
    return Nexus.dynamicOutputProperty({
      name: 'crud',
      typeDefinition: `: NexusPrisma<TypeName, 'crud'>`,
      // FIXME
      // Nexus should improve the type of typeName to be AllOutputTypes
      factory: ({ typeDef: t, typeName, stage }) => {
        if (typeName === GraphQL.rootNames.Subscription) {
          // TODO Lets put a GitHub issue link in this error message
          throw new Error(`t.crud is not yet supported on the 'Subscription' type.`)
        }

        if (typeName !== GraphQL.rootNames.Query && typeName !== GraphQL.rootNames.Mutation) {
          throw new Error(
            `t.crud can only be used on GraphQL root types 'Query' & 'Mutation' but was used on '${typeName}'. Please use 't.model' instead`
          )
        }
        const publishers = getCrudMappedFields(typeName, this.dmmf).reduce((crud, mappedField) => {
          const fieldPublisher: FieldPublisher = (givenConfig) => {
            const inputType = this.dmmf.getInputType(mappedField.field.args[0].inputType.type)
            const publisherConfig = this.buildPublisherConfig({
              field: mappedField.field,
              givenConfig: givenConfig ? givenConfig : {},
            })
            const schemaArgsIndex = indexBy(mappedField.field.args, 'name')

            const originalResolve: GraphQLFieldResolver<any, any, any> = (_root, args, ctx, info) => {
              const prismaClient = this.getPrismaClient(ctx)
              args = transformNullsToUndefined(args, schemaArgsIndex, this.dmmf)
              if (
                typeName === 'Mutation' &&
                (!isEmptyObject(publisherConfig.locallyComputedInputs) ||
                  !isEmptyObject(this.globallyComputedInputs))
              ) {
                args = addComputedInputs({
                  inputType,
                  dmmf: this.dmmf,
                  params: {
                    info,
                    args,
                    ctx,
                  },
                  locallyComputedInputs: publisherConfig.locallyComputedInputs,
                })
              }

              args = this.paginationStrategy.resolve(args)

              return prismaClient[mappedField.prismaClientAccessor][mappedField.operation](args)
            }

            const fieldConfig = this.buildFieldConfig({
              field: mappedField.field,
              publisherConfig,
              typeName,
              operation: mappedField.operation,
              resolve: (root, args, ctx, info) => {
                return givenConfig?.resolve
                  ? givenConfig.resolve(root, args, ctx, info, originalResolve)
                  : originalResolve(root, args, ctx, info)
              },
            })

            if (
              this.assertOutputTypeIsDefined(typeName, mappedField.field.name, publisherConfig.type, stage)
            ) {
              t.field(publisherConfig.alias, fieldConfig)
            }

            this.assertFilteringOrOrderingArgNameExists(
              typeName,
              mappedField.field.outputType.type,
              mappedField.field.name,
              publisherConfig,
              stage
            )

            return crud
          }

          crud[mappedField.field.name] = fieldPublisher

          return crud
        }, {} as PublisherMethods)

        return proxifyPublishers(publishers, typeName, stage, this.options.onUnknownFieldName)
      },
    })
  }

  protected buildDisabledCRUD(): DynamicOutputPropertyDef<'crud'> {
    return Nexus.dynamicOutputProperty({
      name: 'crud',
      factory: () => {
        this.wasCrudUsedButDisabled = true
        return new Proxy(
          {},
          {
            get() {
              return () => {}
            },
          }
        )
      },
    })
  }

  /**
   * Build the `t.model` dynamic output property.
   */
  protected buildModel() {
    return Nexus.dynamicOutputProperty({
      name: 'model',
      typeDefinition: `: NexusPrisma<TypeName, 'model'>`,
      /**
       * This factory implements what .model will actually be.
       *
       * If the user's GQL typedef name matches a PSL model name,
       * then we infer that the user is trying to create a mapping
       * between them. This is the implicit mapping case.
       *
       * Otherwise we need the user to specify what PSL model
       * their GQL object maps to. This is the explicit mapping case.
       *
       * In the implicit case we eagerly do the .model implementation,
       * but in the explicit case we return a function in order that the
       * user may specify the mapping.
       *
       * Examples:
       *
       *    // Given PSL that contains:
       *
       *    model User {
       *      id    String @unique @id @default(uuid())
       *      email String @unique
       *    }
       *
       *    // Example of implicit mapping
       *
       *    objectType({
       *      name: 'User',
       *      definition(t) {
       *        t.model.id()
       *        t.model.email()
       *      }
       *    })
       *
       *    // Example of explicit mapping
       *
       *    objectType({
       *      name: 'Customer',
       *      definition(t) {
       *        t.model('User').id()
       *        t.model('User').email()
       *      }
       *    })
       *
       */
      factory: ({ typeDef, typeName, stage }) => {
        const hasPrismaModel = this.dmmf.hasModel(typeName)
        if (hasPrismaModel) {
          return this.internalBuildModel(typeName, typeDef, stage)
        } else {
          const accessor = (modelName: string) => this.internalBuildModel(modelName, typeDef, stage)

          return proxifyModelFunction(
            accessor,
            typeName,
            stage,
            this.options.onUnknownPrismaModelName,
            this.unknownFieldsByModel
          )
        }
      },
    })
  }

  protected internalBuildModel(
    typeName: string,
    t: Nexus.core.OutputDefinitionBlock<any>,
    stage: Nexus.core.OutputFactoryConfig<any>['stage']
  ) {
    const model = this.dmmf.getModelOrThrow(typeName)
    const outputType = this.dmmf.getOutputType(model.name)

    const publishers = outputType.fields.reduce((acc, field) => {
      const fieldPublisher: FieldPublisher = (givenConfig) => {
        const publisherConfig = this.buildPublisherConfig({
          field,
          givenConfig: givenConfig ?? {},
        })
        if (!this.assertOutputTypeIsDefined(typeName, publisherConfig.alias, publisherConfig.type, stage)) {
          return acc
        }
        this.assertFilteringOrOrderingArgNameExists(
          typeName,
          field.outputType.type,
          publisherConfig.alias,
          publisherConfig,
          stage
        )
        const mapping = this.dmmf.getMapping(typeName)
        const uniqueIdentifiers = Constraints.resolveUniqueIdentifiers(typeName, this.dmmf)
        const schemaArgsIndex = indexBy(field.args, 'name')

        const originalResolve: GraphQLFieldResolver<any, any, any> | undefined =
          field.outputType.kind === 'object'
            ? (root, args, ctx) => {
                const missingIdentifiers = Constraints.findMissingUniqueIdentifiers(root, uniqueIdentifiers)

                if (missingIdentifiers !== null) {
                  throw new Error(
                    `Resolver ${typeName}.${
                      publisherConfig.alias
                    } is missing the following unique identifiers: ${missingIdentifiers.join(', ')}`
                  )
                }

                const prismaClient = this.getPrismaClient(ctx)

                args = transformNullsToUndefined(args, schemaArgsIndex, this.dmmf)
                args = this.paginationStrategy.resolve(args)

                return prismaClient[lowerFirst(mapping.model)]
                  .findOne({
                    where: Constraints.buildWhereUniqueInput(root, uniqueIdentifiers),
                  })
                  [field.name](args)
              }
            : publisherConfig.alias != field.name
            ? (root) => root[field.name]
            : undefined

        const fieldConfig = this.buildFieldConfig({
          field,
          publisherConfig,
          typeName,
          resolve: givenConfig?.resolve
            ? (root, args, ctx, info) => {
                return givenConfig.resolve!(root, args, ctx, info, originalResolve ?? defaultFieldResolver)
              }
            : originalResolve,
        })

        t.field(publisherConfig.alias, fieldConfig)

        return publishers
      }

      acc[field.name] = fieldPublisher
      return acc
    }, {} as PublisherMethods)

    return proxifyPublishers(publishers, typeName, stage, this.options.onUnknownFieldName)
  }

  buildPublisherConfig({
    field,
    givenConfig: { computedInputs, ...otherConfig },
  }: Required<PublisherConfigData>): ResolvedFieldPublisherConfig {
    return {
      pagination: true,
      type: field.outputType.type,
      alias: field.name,
      locallyComputedInputs: computedInputs ? computedInputs : {},
      ...otherConfig,
    }
  }

  buildFieldConfig(config: FieldConfigData): Nexus.core.NexusOutputFieldConfig<any, string> {
    const {
      alias,
      locallyComputedInputs,
      type,
      filtering,
      ordering,
      pagination,
      ...additionalExternalPropsSuchAsPlugins
    } = config.publisherConfig

    return {
      ...additionalExternalPropsSuchAsPlugins,
      type: this.publisher.outputType(config.publisherConfig.type, config.field),
      ...dmmfListFieldTypeToNexus(config.field.outputType),
      args: this.buildArgsFromField(config),
      resolve: config.resolve,
    }
  }

  buildArgsFromField(config: FieldConfigData): Nexus.core.ArgsRecord {
    return this.determineArgs(config).reduce(
      (acc, customArg) => ({
        ...acc,
        [customArg.arg.name]: this.publisher.inputType(customArg) as any,
      }),
      {} as Nexus.core.ArgsRecord
    )
  }

  determineArgs(config: FieldConfigData): CustomInputArg[] {
    if (config.typeName === 'Mutation') {
      return this.argsFromMutationField(config)
    } else if (config.operation === 'findOne') {
      return config.field.args.map((arg) => ({
        arg,
        type: this.dmmf.getInputType(arg.inputType.type),
      }))
    } else {
      return this.argsFromQueryOrModelField(config)
    }
  }

  argsFromMutationField({ publisherConfig, field }: FieldConfigData): CustomInputArg[] {
    return field.args.map((arg) => {
      const prismaClientInputType = this.dmmf.getInputType(arg.inputType.type)
      /*
      Since globallyComputedInputs were already filtered during schema transformation,
      at this point we just need to filter at the resolver-level.
      */
      return {
        arg,
        type: {
          ...prismaClientInputType,
          fields: publisherConfig.locallyComputedInputs
            ? prismaClientInputType.fields.filter(
                (field) => !(field.name in publisherConfig.locallyComputedInputs)
              )
            : prismaClientInputType.fields,
        },
      }
    })
  }

  protected argsFromQueryOrModelField({ typeName, field, publisherConfig }: FieldConfigData) {
    let args: CustomInputArg[] = []

    if (publisherConfig.filtering) {
      const inputObjectTypeDefName = `${field.outputType.type}WhereInput`
      const whereArg = field.args.find(
        (arg) => arg.inputType.type === inputObjectTypeDefName && arg.name === 'where'
      )

      if (!whereArg) {
        throw new Error(`Could not find filtering argument for ${typeName}.${field.name}`)
      }

      const inputType = this.handleInputObjectCustomization(
        publisherConfig.filtering,
        inputObjectTypeDefName,
        field.name,
        typeName
      )

      if (inputType.fields.length > 0) {
        args.push({
          arg: whereArg,
          type: inputType,
        })
      }
    }

    if (publisherConfig.ordering) {
      const orderByTypeName = `${field.outputType.type}OrderByInput`
      const orderByArg = field.args.find(
        (arg) => arg.inputType.type === orderByTypeName && arg.name === 'orderBy'
      )

      if (!orderByArg) {
        throw new Error(`Could not find ordering argument for ${typeName}.${field.name}`)
      }

      const inputType = this.handleInputObjectCustomization(
        publisherConfig.ordering,
        orderByTypeName,
        field.name,
        typeName
      )

      if (inputType.fields.length > 0) {
        args.push({
          arg: orderByArg,
          type: inputType,
        })
      }
    }

    if (publisherConfig.pagination) {
      const paginationsArgs =
        publisherConfig.pagination === true
          ? field.args.filter((a) => {
              const argNames = this.paginationStrategy.paginationArgNames as string[]
              return argNames.includes(a.name)
            })
          : field.args.filter((arg) => {
              return (publisherConfig.pagination as any)[arg.name] === true
            })

      args.push(
        ...paginationsArgs.map((a) => {
          if (a.inputType.kind === 'scalar' || a.inputType.kind === 'enum') {
            return {
              arg: a,
              type: { name: a.inputType.type },
            }
          } else {
            return {
              arg: a,
              type: this.dmmf.inputTypesIndex[a.inputType.type],
            }
          }
        })
      )
    }

    return args
  }

  /**
   * This handles "tailored field feature publishing".
   *
   * With tailored field feature publishing, users can specify that only some
   * fields of the PSL model are exposed under the given field feature. For
   * example, in the following...
   *
   * ```ts
   * t.model.friends({ filtering: { firstName: true, location: true } })
   * ```
   *
   * ...the field feature is "filtering" and the user has tailored it so that
   * only "firstName" and "location" of the field's type (e.g. "User") are
   * exposed to filtering on this field. So the resulting GQL TypeDef would look
   * something like:
   *
   * ```ts
   * ...
   * friends(where: { firstName: ..., location: ..., }): [User]
   * ...
   * ```
   */
  protected handleInputObjectCustomization(
    fieldWhitelist: Record<string, boolean> | boolean,
    inputTypeName: string,
    fieldName: string,
    graphQLTypeName: string
  ): DmmfTypes.InputType {
    const prismaClientObject = this.dmmf.getInputType(inputTypeName)

    // If the publishing for this field feature (filtering, ordering, ...)
    // has not been tailored then we may simply pass through the backing
    // version as-is.
    //
    if (fieldWhitelist === true) {
      return prismaClientObject
    }

    // REFACTOR use an intersection function
    const whitelistedFieldNames = Object.keys(fieldWhitelist)
    const userExposedObjectFields = prismaClientObject.fields.filter((field) =>
      whitelistedFieldNames.includes(field.name)
    )

    const uniqueName = prismaClientObject.isWhereType
      ? this.argsNamingStrategy.whereInput(graphQLTypeName, fieldName)
      : this.argsNamingStrategy.orderByInput(graphQLTypeName, fieldName)

    return {
      ...prismaClientObject,
      name: uniqueName,
      fields: userExposedObjectFields,
    }
  }

  protected assertOutputTypeIsDefined(
    typeName: string,
    fieldName: string,
    outputType: string,
    stage: 'walk' | 'build'
  ): boolean {
    if (
      this.options.nexusBuilder.hasType(outputType) ||
      GraphQL.isScalarType(outputType) || // scalar types are auto-published
      !this.dmmf.hasModel(outputType) // output types that are not models are auto-published
    ) {
      return true
    }

    const message = `Your GraphQL \`${typeName}\` object definition is projecting a field \`${fieldName}\` with \`${outputType}\` as output type, but \`${outputType}\` is not defined in your GraphQL Schema`

    raiseErrorOrTriggerHook(
      this.options.onUnknownFieldType,
      {
        unknownFieldType: outputType,
        typeName,
        fieldName,
        error: new Error(message),
      },
      message,
      stage
    )

    return false
  }

  protected assertArgNameExists(
    parentTypeName: string,
    prismaOutputTypeName: string,
    fieldName: string,
    config: FieldPublisherConfig,
    stage: 'build' | 'walk',
    configProperty: 'filtering' | 'ordering'
  ): { wrongArgNames: string[] } | true {
    if (!config[configProperty] || config[configProperty] === true) {
      return true
    }

    const argNames = Object.keys(config[configProperty] as Record<string, boolean>)
    const typeNameFieldNames = this.dmmf.getModelOrThrow(prismaOutputTypeName).fields.map((f) => f.name)
    const wrongArgNames = argNames.filter(
      (filteringFieldName) => !typeNameFieldNames.includes(filteringFieldName)
    )

    if (wrongArgNames.length === 0) {
      return true
    }

    const actionWord = configProperty === 'filtering' ? 'filter' : 'order'
    const renderMessage = (argName: string) =>
      `Your GraphQL \`${parentTypeName}\` object definition is projecting a relational field \`${fieldName}\`. On it, you are declaring that clients be able to ${actionWord} by Prisma \`${prismaOutputTypeName}\` model field \`${argName}\`. However, your Prisma model \`${prismaOutputTypeName}\` model has no such field \`${argName}\``

    const message = wrongArgNames.map((argName) => renderMessage(argName)).join('\n')

    raiseErrorOrTriggerHook(
      this.options.onUnknownArgName,
      {
        unknownArgNames: wrongArgNames,
        typeName: prismaOutputTypeName,
        fieldName,
        error: new Error(message),
      },
      message,
      stage
    )

    return { wrongArgNames }
  }

  protected assertFilteringOrOrderingArgNameExists(
    parentTypeName: string,
    prismaOutputTypeName: string,
    fieldName: string,
    config: FieldPublisherConfig,
    stage: 'build' | 'walk'
  ): void {
    this.assertArgNameExists(parentTypeName, prismaOutputTypeName, fieldName, config, stage, 'filtering')
    this.assertArgNameExists(parentTypeName, prismaOutputTypeName, fieldName, config, stage, 'ordering')
  }
}
