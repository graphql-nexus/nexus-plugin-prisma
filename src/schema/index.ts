import { plugin } from '@nexus/schema'
import { build as buildNexusPrismaTypes, Options, InternalPublicOptions } from './builder'
import { CRUD_NOT_ENABLED } from './warnings'

/**
 * Create a nexus-plugin-prisma instance to be passed into the Nexus plugins array.
 *
 * The nexus plugin extends the Nexus DSL with `t.model` and `t.crud`. These
 * allow you to expose data from your Prisma layer and operations against that
 * data. Once this plugin is installed, you can do things such as:
 *
 * ```ts
 * objectType({
 *   name: 'User',
 *   definition(t) {
 *     t.model.id()
 *     t.model.email()
 *   }
 * })
 *
 * queryType({
 *   definition (t) {
 *      t.crud.user()
 *      t.crud.users({ filtering: true, ordering: true })
 *    }
 * })
 * ```
 *
 * You must ensure the Prisma Client JS has been generated prior as it provides a
 * data representation of the available models and CRUD operations against them.
 *
 * Like Nexus, NexusPrisma has its own Typegen. By default It will be generated
 * synchronously when Nexus runs the plugin, _if_ `process.env.NODE_ENV` is
 * `undefined` or `"development"`. Typegen can be explicitly enabled or disabled
 * via the shouldGenerateArtifacts option. This mirrors Nexus' own typegen
 * approach. This system will change once the Nexus plugin system has first
 * class support for typegen.
 */
export function nexusSchemaPrisma(options?: Options) {
  const allOptions: InternalPublicOptions = options ?? {}
  let wasCrudUsedButDisabled: null | (() => boolean) = null

  return plugin({
    name: 'nexus-plugin-prisma',
    onInstall: (nexusBuilder) => {
      const { types, wasCrudUsedButDisabled: wasCrudUsed } = buildNexusPrismaTypes({
        ...allOptions,
        nexusBuilder,
      })

      wasCrudUsedButDisabled = wasCrudUsed

      return {
        types,
      }
    },
    onBeforeBuild() {
      if (wasCrudUsedButDisabled?.() === true) {
        if (allOptions.initializedByFramework === true) {
          CRUD_NOT_ENABLED.framework()
        } else {
          CRUD_NOT_ENABLED.schema()
        }
      }
    },
  })
}

export { Options } from './builder'
