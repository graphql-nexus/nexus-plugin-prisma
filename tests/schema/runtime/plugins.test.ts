import { fieldAuthorizePlugin, objectType } from '@nexus/schema'
import { createRuntimeTestContext } from '../__client-test-context'

let ctx = createRuntimeTestContext()

it('forwards plugins to t.model', async () => {
  const datamodel = `  
datasource db {
  provider = "postgresql"
  url      = "postgresql://"
}

model User {
  id    Int     @id @default(autoincrement())
  name  String
}
  `
  const types = [
    objectType({
      name: 'User',
      definition(t: any) {
        t.model.id({
          authorize() {
            return new Error('nope')
          },
        })
      },
    }),
    objectType({
      name: 'Query',
      definition(t: any) {
        t.crud.findManyUser()
      },
    }),
  ]

  const { graphqlClient, dbClient } = await ctx.getContext({
    datamodel,
    types,
    plugins: [fieldAuthorizePlugin()],
  })

  await dbClient.user.create({
    data: {
      name: 'Foo',
    },
  })

  try {
    await graphqlClient.request(`{
      users {
        id
      }
    }`)
  } catch (e) {
    expect(e).toMatchSnapshot()
  }
})

it('forwards plugins to t.crud', async () => {
  const datamodel = `
    model User {
      id    Int     @id @default(autoincrement())
      name  String
    }
  `
  const User = objectType({
    name: 'User',
    definition(t: any) {
      t.model.id()
    },
  })

  const Query = objectType({
    name: 'Query',
    definition(t: any) {
      t.crud.findManyUser({
        authorize() {
          return new Error('nope')
        },
      })
    },
  })

  const { graphqlClient, dbClient } = await ctx.getContext({
    datamodel,
    types: [Query, User],
    plugins: [fieldAuthorizePlugin()],
  })

  await dbClient.user.create({
    data: {
      name: 'Foo',
    },
  })

  try {
    await graphqlClient.request(`{
      users {
        id
      }
    }`)
  } catch (e) {
    expect(e).toMatchSnapshot()
  }
})
