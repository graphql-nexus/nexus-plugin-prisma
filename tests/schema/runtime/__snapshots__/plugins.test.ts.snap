// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`forwards plugins to t.crud 1`] = `[Error: Not authorized: {"response":{"data":null,"errors":[{"message":"Not authorized","locations":[{"line":2,"column":7}],"path":["users"]}],"status":200},"request":{"query":"{\\n      users {\\n        id\\n      }\\n    }"}}]`;

exports[`forwards plugins to t.model 1`] = `[Error: Not authorized: {"response":{"data":null,"errors":[{"message":"Not authorized","locations":[{"line":3,"column":9}],"path":["users",0,"id"]}],"status":200},"request":{"query":"{\\n      users {\\n        id\\n      }\\n    }"}}]`;
